lib/image_merger.dart

```diff
@@ -0,0 +1,142 @@
+import 'dart:async';
+import 'dart:typed_data';
+import 'dart:ui';
+
+import 'extensions/vector2.dart';
+import 'flame.dart';
+
+class ImageMerger {
+  /// The images that are going to be merged together.
+  final List<Image> _images = [];
+
+  /// The positions where each [_images] will be added.
+  final List<Vector2> _positions = [];
+
+  /// The sources that should be used on each [_images].
+  final List<Rect> _sources = [];
+
+  /// Add an image to the [ImageMerger].
+  ///
+  /// The [image] will be added at the given [position] on the generated image.
+  ///
+  /// An optional [source] can be used to only add the data that is within the
+  /// [source] of the [image].
+  /// TODO(wfln): Add [BlendMode support].
+  void add(
+    Image image,
+    Vector2 position, {
+    Rect source,
+  }) {
+    assert(image != null, 'Image is required to add to the Atlas');
+    assert(position != null, 'Position is required');
+    assert(
+      position.x >= 0 && position.y >= 0,
+      'Position values have to be equal or greater than zero',
+    );
+    assert(
+      source == null ||
+          source.width <= image.width &&
+              source.height <= image.height &&
+              source.top + source.height <= image.height &&
+              source.top >= 0 &&
+              source.left + source.width <= image.width &&
+              source.left >= 0,
+      'Source rect should fit within in the image constraints',
+    );
+    source ??= Rect.fromLTWH(
+      0,
+      0,
+      image.width.toDouble(),
+      image.height.toDouble(),
+    );
+
+    _images.add(image);
+    _positions.add(position);
+    _sources.add(source);
+  }
+
+  /// Merge all the images into a single atlas.
+  Future<Image> merge() async {
+    var output = const Rect.fromLTWH(0, 0, 0, 0);
+    var data = Uint8List(_toLength(Vector2.zero()));
+
+    for (var i = 0; i < _images.length; i++) {
+      final image = _images[i];
+      final position = _positions[i];
+      final source = _sources[i];
+
+      final destination = Rect.fromLTWH(
+        position.x,
+        position.y,
+        source.width,
+        source.height,
+      );
+
+      // If the rect is not fully contained inside the current outputRect,
+      // we have to resize the data list.
+      if (!(output.contains(destination.topLeft) &&
+          output.contains(destination.topRight) &&
+          output.contains(destination.bottomLeft) &&
+          output.contains(destination.bottomRight))) {
+        final newOutput = output.expandToInclude(destination);
+        final newData = Uint8List(
```

Comment:
Last time I needed to concatenate images I used PictureRecorder, which was way easier than manipulating the bytes themselves, any reason for you to have used bytes directly?

Reply:
I originally wrote this as a command line tool, I already refactored it to the PictureRecorder in my local branch👌🏻